scrapbook/managers.py

```diff
@@ -0,0 +1,446 @@
+        return self._managers.__len__()
+
+    def register(self, encoder_name, store_name, manager):
+        """
+        Registers a new name to a particular encoder
+
+        Parameters
+        ----------
+        encoder_name: str (or None)
+            Name of the mime subtype parsed by the encoder. None implies any.
+        store_name: str (or None)
+            The specific storage type that this encoder supports. None implies any.
+        encoder: obj
+            The object which implements the required encoding functions.
+        """
+        # Add the registered name so we can reverse look-up these fields
+        manager.encoder_name = encoder_name
+        manager.store_name = store_name
+        self[(encoder_name, store_name)] = manager
+
+    def register_encoder(self, name, encoder):
+        self.register(name, None, encoder)
+
+    def register_store(self, name, store):
+        self.register(None, name, store)
+
+    def deregister(self, encoder_name, store_name):
+        """
+        Removes a particular encoder from the registry
+
+        Parameters
+        ----------
+        name: str
+            Name of the mime subtype parsed by the encoder.
+        store: str (optional)
+            The specific storage type that this encoder supports.
+        """
+        del self[(encoder_name, store_name)]
+
+    def reset(self):
+        """
+        Resets the registry to have no encoders.
+        """
+        self._managers = OrderedDict()
+
+    def find_manager(self, scrap):
+        if scrap.encoder is not None and scrap.store is not None:
+            return self.get((scrap.encoder, scrap.store))
+
+        for manager in reversed(list(self.values())):
+            # Only check for managers here that can both encode and store
+            if not (manager.encoder_name and manager.store_name):
+                continue
+            try:
+                # Find matches both the encoder and the store are valid for the
+                # constraints the scrap provides
+                if (
+                    scrap.encoder == manager.encoder_name
+                    or (scrap.encoder is None and manager.encodable(scrap))
+                ) and (
+                    scrap.store == manager.store_name
+                    or (scrap.store is None and manager.storable(scrap))
+                ):
+                    return manager
+            except AttributeError:
+                # Skip any manager that don't implement encoding and storage
+                pass
+
+    def find_encoder(self, scrap):
+        """
+        Searches for encoders in order of registration that can encode a given
+        scap.
+
+        Parameters
+        ----------
+        scrap: Scrap
+            A partially filled in scrap with data that needs decoding
+        """
+        if scrap.encoder is not None:
+            return self.get((scrap.encoder, None))
+
+        for encoder in reversed(list(self.values())):
```

Comment:
Curious: is there a reason you reverse the list of encoders here and elsewhere? Does it have something to do with the properties of an `OrderedDict`?

Reply:
Yes. You can't prefix to OrderedDicts easily, which we need to be able to do, it requires rebuilding the whole structure.